sound/soc/soc-pcm.c

```diff
@@ -3150,6 +3152,44 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 		ret = snd_pcm_new_internal(rtd->card->snd_card, new_name, num,
 				playback, capture, &pcm);
 	} else {
+
+		/*
+		 * for virtual FE dai links, there is no need
+		 * to register PCM device. So only allocate memory for
+		 * pcm device and substream for the requested direction
+		 */
+		if (rtd->dai_link->virtual) {
+			struct snd_pcm_str *pstr;
```

Comment:
I would offload this logic as a static function.

Reply:
and can any part of this be done by call core ALSA functions directly ?

Reply:
The only function that does this is snd_new_pcm() but that also registers the pcm device. So I created this to avoid that part.